### PR TITLE
Fix `RCTConvert UIFont` loosing styles

### PR DESCRIPTION
**Pull request by grabbou**
This fixes odd issue #2140 that I think might have been introduced back in the day when `System` font had landed.

**tl;dr** This PR makes sure we properly recognise `System` font when merging with other properties so it does not loose styles.

The problem here (and also reason why we have missed it so far) is that most of the codebase calls:
```objc
+ (UIFont *)UIFont:(UIFont *)font withFamily:(id)family
              size:(id)size weight:(id)weight style:(id)style
   scaleMultiplier:(CGFloat)scaleMultiplier
```
or 
```objc
+ (UIFont *)UIFont:(id)json
```
directly. 

That means we just generate a font from arbitrary `json` string and apply it to our views. That works and that's the part we have fully covered. However, we forgot to test these:

```objc
+ (UIFont *)UIFont:(UIFont *)font withWeight:(id)weight
```
and similar - and unfortunately these were failing. The reason it's been reported as a `TextInput` bug is that if I am not wrong - we use it only there.

**Background**

`TextInput` exports custom properties [here](https://github.com/facebook/react-native/blob/master/Libraries/Text/RCTTextFieldManager.m#L100) and calls the functions mentioned above to apply styles to the current font.

We have two cases here:
1) First time these are called with a default font set during initialisation that has a real system fontFamily (in my case it's `.SF-UI-Text`). That means it's not going to be recognised as a system font since we only assume `System` string is valid. They will go to the custom-font code which for some reason will strip all styles but fontSize (not sure why, it's a bit complicated part to follow code-wise, but I added relevant test cases to make sure custom font families also work, e.g. `foobar` and `Arial`)
2) When font is created from an arbitrary json - we are returning System font on iOS8+ since that's our default behaviour in such case. However - the returned font does not have `System` fontFamily, but a real device fontFamily, which makes us go to point `#1` again.

Fix is to not only compare the `System` string as a fontFamily, but also retrieve real system font family for given weight and size (note this can change depending on fontSize, related PR: #7044).

Reproductions & screen confirming `all good` can be found in the linked issue.

### CODE REVIEW

File: React/Base/RCTConvert.m [hunk 1]
```diff
@@ -644,15 +655,9 @@ + (UIFont *)UIFont:(UIFont *)font withFamily:(id)family
...
-      // Not a valid font or family
+      // Not a valid font or family, fallback with system one
       RCTLogError(@"Unrecognized font family '%@'", familyName);
-      if ([UIFont respondsToSelector:@selector(systemFontOfSize:weight:)]) {
-        font = [UIFont systemFontOfSize:fontSize weight:fontWeight];
-      } else if (fontWeight > UIFontWeightRegular) {
-        font = [UIFont boldSystemFontOfSize:fontSize];
-      } else {
-        font = [UIFont systemFontOfSize:fontSize];
-      }
```
Comment (grabbou) [inline React/Base/RCTConvert.m]: My commit msg is wrong actually - I assumed `systemFontOfSize` was iOS8+ which in fact only applies to `systemFontOfSize:weight`. However I still think it might be easier to just call the method again without fontFamily.

File: React/Base/RCTConvert.m [hunk 2]
```diff
@@ -582,6 +583,7 @@ + (UIFont *)UIFont:(UIFont *)font withFamily:(id)family
   NSString *const RCTIOS8SystemFontFamily = @"Helvetica Neue";
   const RCTFontWeight RCTDefaultFontWeight = UIFontWeightRegular;
   const CGFloat RCTDefaultFontSize = 14;
+  const bool HasSystemFontWeightSelector = [UIFont respondsToSelector:@selector(systemFontOfSize:weight:)];
```
Comment (alloy) [inline React/Base/RCTConvert.m]: Should probably be a `BOOL`.

File: React/Base/RCTConvert.m [hunk 3]
```diff
@@ -607,10 +610,18 @@ + (UIFont *)UIFont:(UIFont *)font withFamily:(id)family
   isItalic = style ? [self RCTFontStyle:style] : isItalic;
   fontWeight = weight ? [self RCTFontWeight:weight] : fontWeight;
 
+  // Font is a systemFont if its familyName is either equal to `System` or
+  // real system font family determined by systemFontOfSize selector
+  bool isSystemFont = [familyName isEqual:RCTDefaultFontFamily];
```
Comment (alloy) [inline React/Base/RCTConvert.m]: Same here (`BOOL`).

**Comment by grabbou** [comment]
@brentvatne after this PR, you've changed quite a few things in RCTConvert. Would you mind checking if your diff did fix similar things? If not, I think it would be cool to rebase, add these missing tests and ship it.

**Comment by javache** [comment]
I've got a similar fix to this up for review inside, I'll link it in #2140. Please open a new PR if there's still things left to be addressed.
